src/backend/distributed/executor/multi_executor.c

```diff
@@ -691,6 +693,82 @@ SetLocalMultiShardModifyModeToSequential()
 }
 
 
+void
+EnsureSequentialMode(ObjectType objType)
+{
+	char *objTypeString = GetObjectTypeString(objType);
+
+	if (ParallelQueryExecutedInTransaction())
+	{
+		ereport(ERROR, (errmsg("cannot run %s command because there was a "
+							   "parallel operation on a distributed table in the "
+							   "transaction", objTypeString),
+						errdetail("When running command on/for a distributed %s, Citus "
+								  "needs to perform all operations over a single "
+								  "connection per node to ensure consistency.",
+								  objTypeString),
+						errhint("Try re-running the transaction with "
+								"\"SET LOCAL citus.multi_shard_modify_mode TO "
+								"\'sequential\';\"")));
+	}
+
+	ereport(DEBUG1, (errmsg("switching to sequential query execution mode"),
+					 errdetail(
+						 "A command for a distributed %s is run. To make sure subsequent "
+						 "commands see the %s correctly we need to make sure to "
+						 "use only one connection for all future commands",
+						 objTypeString, objTypeString)));
+
+	SetLocalMultiShardModifyModeToSequential();
+}
+
+
+static char *
+GetObjectTypeString(ObjectType objType)
+{
+	switch (objType)
+	{
+		case OBJECT_COLLATION:
+		{
+			return "collation";
+		}
+
+		case OBJECT_DATABASE:
+		{
+			return "database";
+		}
+
+		case OBJECT_EXTENSION:
+		{
+			return "extension";
+		}
+
+		case OBJECT_FUNCTION:
+		{
+			return "function";
+		}
+
+		case OBJECT_SCHEMA:
+		{
+			return "schema";
+		}
+
+		case OBJECT_TYPE:
+		{
+			return "type";
+		}
+
+		default:
+		{
+			ereport(DEBUG1, (errmsg("unsupported object type"),
+							 errdetail("Please add string conversion for the object.")));
+		}
```

Comment:
at first glance I thought this was an error, instead it's a debug message, which at least keeps the original error.
To make the code more understandable, could we move the return statement from outside the switch here?

Lastly did you make a consious decission to return an empty string? I think errors emitting an empty string look funky and are sometimes not super descriptive. Maybe a catchall like `"object"` would make the errors slightly more readable, just speculating here.

Whatever you go with, an extra comment wouldn't hurt.